fix: parse error when only outputs is changed

taro-kayo opened the pull request:
Close #117

Related PR: https://github.com/mercari/tfnotify/pull/109

taro-kayo commented on pkg/terraform/parser.go:
```diff
@@ -150,8 +150,8 @@ func (p *PlanParser) Parse(body string) ParseResult { //nolint:cyclop
 		}
-		if startChangeOutput != -1 && endChangeOutput == -1 && strings.HasPrefix(line, "Plan: ") { // https://github.com/hashicorp/terraform/blob/dfc12a6a9e1cff323829026d51873c1b80200757/internal/command/views/plan.go#L306
-			endChangeOutput = i + 1
+		if startChangeOutput != -1 && endChangeOutput == -1 && strings.HasPrefix(line, "-----") { // https://github.com/hashicorp/terraform/blob/1ac7a37d00f3c796f816070847bf02109cb9cab2/internal/command/views/operation.go#L142
```
To include changes to outputs coming after `Plan:`, we have to wait for a horizontal rule which comes after `Changes to Outputs:`.

taro-kayo commented on pkg/terraform/parser.go:
```diff
@@ -201,6 +201,10 @@ func (p *PlanParser) Parse(body string) ParseResult { //nolint:cyclop
 
 	changeResult := ""
 	if startChangeOutput != -1 {
+		// if we get here before finding a horizontal rule, output all remaining.
+		if endChangeOutput == -1 {
+			endChangeOutput = len(lines) - 1
+		}
```
If [`TF_IN_AUTOMATION` environment variable](https://developer.hashicorp.com/terraform/tutorials/automation/automate-terraform#controlling-terraform-output-in-automation) is set, no horizontal rules appear after `Terraform will perform the following actions:`. 😢

taro-kayo commented on pkg/terraform/parser_test.go:
```diff
@@ -351,6 +455,56 @@ func TestPlanParserParse(t *testing.T) {
+			name: "plan output changes only pattern 0.12",
+			body: planOnlyOutputChangesSuccessResult0_12,
+			result: ParseResult{
+				Result:             "Plan: 0 to add, 0 to change, 0 to destroy.",
```
This is not ideal, but it's very older version... 🙈

suzuki-shunsuke commented:
Thank you for your contribution!

suzuki-shunsuke commented:
I'm reviewing this pr yet, but I found an issue.

```console
$ terraform version
Terraform v1.5.7
on darwin_arm64
+ provider registry.terraform.io/hashicorp/github v5.35.0
```

main.tf

```tf
cat main.tf 
import {
  id = "tfaction"
  to = github_repository.tfaction
}

output "foo" {
  value = "foo"
}
```

```console
$ tfcmt plan -- terraform plan

Changes to Outputs:
  + foo = "foo"

You can apply this plan to save these new output values to the Terraform
state, without changing any real infrastructure.
╷
│ Error: Import block target does not exist
│ 
│   on main.tf line 1:
│    1: import {
│ 
│ The target for the given import block does not exist. If you wish to
│ automatically generate config for this resource, use the
│ -generate-config-out option within terraform plan. Otherwise, make sure the
│ target resource exists within your configuration. For example:
│ 
│   terraform plan -generate-config-out=generated.tf
╵
```

AS IS (tfcmt v4.5.0)

<img width="955" alt="image" src="https://github.com/suzuki-shunsuke/tfcmt/assets/13323303/c4f50d7a-0b48-427d-b4e4-f75513d56342">

TO BE

<img width="955" alt="image" src="https://github.com/suzuki-shunsuke/tfcmt/assets/13323303/e2b80ef6-9f69-4e07-9aa3-358b1122fb1b">

At least, the output `Changes to Outputs:` doesn't necessarily mean the success of `terraform plan`.

taro-kayo commented:
Thanks for testing. 😃 
I'll take a look.

taro-kayo commented:
OK, I tested.
So, it seems that the behavior in case of error is not this PR problem.
There may be room for improvement regarding the fact that `Click me` is not displayed in the case of `no error, resource no changes, outputs changing`. 😃  

## failed, resource changing, outputs changing

Both v4.5.0 and this PR version have the same result.

> ![Screenshot 2023-09-09 at 14 32 29](https://github.com/suzuki-shunsuke/tfcmt/assets/40859826/179cb1ba-c29e-4c9f-80ab-ab66facdd7bf)

## failed, resource changing, outputs no changes

Same as `failed, resource changing, outputs changing` case above.

## failed, resource no changes, outputs changing

### v4.5.0

> ![Screenshot 2023-09-09 at 14 37 00](https://github.com/suzuki-shunsuke/tfcmt/assets/40859826/c7d6b8c9-632c-4f56-a6ff-0818c6bd8503)

### this PR version

> ![Screenshot 2023-09-09 at 14 37 42](https://github.com/suzuki-shunsuke/tfcmt/assets/40859826/3d4ccf8e-62d8-41a3-a859-550cf1205090)

## no error, resource changing, outputs changing

### v4.5.0

> ![Screenshot 2023-09-09 at 14 46 30](https://github.com/suzuki-shunsuke/tfcmt/assets/40859826/1bf1ed44-857d-41b0-a178-83da7d4ac9ee)

### this PR version

> ![Screenshot 2023-09-09 at 14 46 45](https://github.com/suzuki-shunsuke/tfcmt/assets/40859826/d5d168de-bd4b-4013-9985-605211b84126)

## no error, resource changing, outputs no changes

Both v4.5.0 and this PR version have the same result.

> ![Screenshot 2023-09-09 at 14 42 35](https://github.com/suzuki-shunsuke/tfcmt/assets/40859826/06524274-1ad7-4835-bd1e-74394f2fe298)

## no error, resource no changes, outputs changing

### v4.5.0

> ![Screenshot 2023-09-09 at 14 40 45](https://github.com/suzuki-shunsuke/tfcmt/assets/40859826/ec779b65-2730-4ee0-9d2f-72e386460f89)

### this PR version

> ![Screenshot 2023-09-09 at 14 40 55](https://github.com/suzuki-shunsuke/tfcmt/assets/40859826/1fa64669-4d8d-4fa3-b893-35dc59accc1e)

taro-kayo commented:
Fixed the problem I noticed in images in [the comment above](https://github.com/suzuki-shunsuke/tfcmt/pull/896#issuecomment-1712426846). 🎉 

Problem:  `Note: ...` is included in `Change Result`.
fixed: https://github.com/suzuki-shunsuke/tfcmt/pull/896/commits/e40fa523a3d68eaeb5e554b534fb46f22a06173b

## before

> ![Screenshot 2023-09-09 at 16 17 37](https://github.com/suzuki-shunsuke/tfcmt/assets/40859826/5610efc6-3a6d-46b5-a310-1b50683f5f82)

## after

> ![Screenshot 2023-09-09 at 16 18 11](https://github.com/suzuki-shunsuke/tfcmt/assets/40859826/c13f82e3-322d-48d3-b163-5e560eb0b2fa)

---

Problem: `Change Result` is not displayed when only tfstate outputs is changed.
fixed: https://github.com/suzuki-shunsuke/tfcmt/pull/896/commits/5d98cbaddc2756465bc3535a43b5000d8449ce1f

I copied the test cases from https://github.com/mercari/tfnotify/pull/109, but it may be that Terraform's output specification changed at some point.

## before

> ![Screenshot 2023-09-09 at 16 19 54](https://github.com/suzuki-shunsuke/tfcmt/assets/40859826/091bec4a-938d-4884-80f9-1a4f789fdf3d)


## after

> ![Screenshot 2023-09-09 at 16 20 14](https://github.com/suzuki-shunsuke/tfcmt/assets/40859826/d5a59007-0fe5-4466-a5dd-20aeca2847fb)

suzuki-shunsuke commented:
Thank you for your update.
I think the meaning of the result `Changes to Outputs` isn't clear for users.

<img width="238" alt="image" src="https://github.com/suzuki-shunsuke/tfcmt/assets/13323303/5bf15e93-7da1-4242-b877-2a6092dcad98">

I'd like to make this message easier to understand.

For example, how about `No resource changes, but Outputs are changed` or `Only Outputs are changed`?

### Plan Result

```
No resource changes, but Outputs are changed
```

### Plan Result

```
Only Outputs are changed
```

taro-kayo commented:
That's what I thought too, but pretended not to see. 🙈 

Since it is plan, how about changing it to `Only output will be changed.`?
I'll fix it if this is okay.

suzuki-shunsuke commented:
> Since it is plan, how about changing it to `Only output will be changed.`?

Oh, I see.
Looks good.

Indeed, Terraform uses `will`.

e.g.

> Terraform will perform the following actions:

> null_resource.bar will be created

Could you change to `Only Outputs will be changed.`?

- I'd like to capitalize `Outputs` because this `Outputs` is not a general word but a Terraform specific feature
  - Terraform also capitalize `Outputs`: e.g. `Changes to Outputs`
- I'd like to make `Outputs` plural

suzuki-shunsuke commented:
Other than that it looks good.

taro-kayo commented:
`Outputs` looks better. 👍 

fixed: https://github.com/suzuki-shunsuke/tfcmt/pull/896/commits/6fa879f84ecbd3f819343e57b4618db5fbafab78

I'm not too sure about this fix. 😅

suzuki-shunsuke commented:
Thank you for your great work!

suzuki-shunsuke commented:
The prerelease version v4.5.1-1 is out.
https://github.com/suzuki-shunsuke/tfcmt/releases/tag/v4.5.1-1

We'll verify this a little, and if there is no problem, we'll release v4.5.1.

suzuki-shunsuke commented:
v4.5.1 is out 🎉
https://github.com/suzuki-shunsuke/tfcmt/releases/tag/v4.5.1

Thank you for your contribution!
